src/rgw/rgw_rados.cc

```diff
@@ -8281,6 +8281,17 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
     inline bool at_end() const {
       return cursor == end;
     }
+    // add the next unique candidate, or return false if we reach the end
+    bool next_candidate(std::map<std::string, size_t>& candidates,
+                        size_t tracker_idx) {
+      while (!at_end()) {
+        if (candidates.emplace(entry_name(), tracker_idx).second) {
+          return true;
+        }
+        advance(); // skip duplicate common prefixes
+      }
+      return false;
+    }
```

Comment:
Not sure I like the organization of this. This is logic from the outside pushed into the struct, which is realy only a tracker (a fancy iterator), and doesn't need to know about candidates and its index into that collection. I would prefer to maintain the separation of concerns.